unknown_type/[M]378.cpp

```diff
@@ -0,0 +1,63 @@
+// Problem: 378. Kth Smallest Element in a Sorted Matrix - Medium
+// Link: https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/
+
+/* Nhận xét:
+- Bài này anh chịu rồi :))
+- Bài này sau khi sửa theo C2 thì thời gian chạy thậm chỉ còn tăng lên, có vẻ như overhead là quá lớn. (C1: 32ms vs C2: 52ms). Chú nghĩ sao?
+*/
+
+// C1: Sort toàn bộ dãy rồi chọn phần tử thứ k
+class Solution {
+public:
+    int kthSmallest(vector<vector<int>>& matrix, int k) {
+        vector<int> a;
+        for (auto &vt : matrix) {
+            for (auto &val : vt) {
+                a.push_back(val);
+            }
+        }
+        sort(a.begin(), a.end());
+        return a[k-1];
+    }
+};
+
+// C2: Bắt đầu từ phần tử bé nhất của mỗi hàng, sau đó thêm vào heap, chọn lần lượt cho tới đến khi đủ k phần tử, mỗi khi loại một phần tử ra thì chỉ cần thêm vào 
+// phần tử tiếp theo của hàng đó
+// => Time complexity: O(nlog(n) + klog(n)). nlogn cho thao tác tạo heap ban đầu. Sau đó, k lần lấy phần tử ra, mỗi lần lại thêm vào 1 phần tử nên số phần tử
+// trong heap luôn được giữ là n => thao tác push heap sẽ có độ phức tạp log(n)
+class Solution {
+public:
+    typedef pair<int, pair<int, int>> pp;
+    
+    class Comp {
+    public:
+        bool operator() (const pp &a, const pp &b) {
+            return a.first > b.first;
+        }
+    };
+    
+    int kthSmallest(vector<vector<int>>& matrix, int k) {
+        int n = matrix.size();
+        
+        priority_queue<pp, vector<pp>, Comp> heap;
```

Comment:
Chú thử bỏ cái tham số thứ 2 (cái vector<pp> kia đi xem sao)

Reply:
Để khai báo kiểu này thì tham số đó là bắt buộc. Anh thử dùng một kiểu khai báo khác rồi, chú kiểm tra lại xem thế nào.
Nói chung gần như không khác gì nhau với thời gian chạy thì vẫn thế.

Reply:
ok cố gắng set cho mình cái code chuẩn này khi code priority queue.  Thực tế bộ test yếu thì nó mới chạy chậm hơn cái cách trâu kia thôi. Average nó sẽ nhanh hơn.

Reply:
Nên dùng cách này: priority_queue heap(comp, vector<pp>());
thay cho cách này: priority_queue<pp, vector<pp>, Comp> heap; 
ah?

Reply:
ý a là dùng Priority queue thay cho cái heap ngày trươc chú dùng. Còn 2 cách khởi tạo kia chắc tương đương nhau. Ko ảnh hưởng

Reply:
Thật ra thì 2 cái nó cũng gần tương tự nhau đấy chú, chẳng qua cái này nó gói vào 1 lớp đàng hoàng nên trông code sẽ gọn và chuyên nghiệp hơn. 
Như ở trong link document này có nói: https://www.cplusplus.com/reference/queue/priority_queue/

Support of random access iterators is required to keep a heap structure internally at all times. This is done automatically by the container adaptor by automatically calling the algorithm functions make_heap, push_heap and pop_heap when needed.

Reply:
OK chú. Vậy cứ keep cái cách code sạch đẹp này cho a.

Reply:
Thanks bro!